File: src/providers/openai-realtime/rules/python/buffer_audio_until_session_ready.py
"""openai-realtime-buffer-audio-until-session-ready

Parity with the JS rule: audio that arrives before the Realtime socket
reaches its open/ready state must be queued and flushed once open, not
dropped by an early-return branch.
"""
from __future__ import annotations

import ast
import re
import sys
from pathlib import Path

from ast_utils import loc

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

RULE_KEY = "openai-realtime-buffer-audio-until-session-ready"

STATE_ATTR_NAMES = frozenset({"state", "readyState", "ready_state"})
FLAG_NAME_RE = re.compile(r"open|ready|connected", re.I)
QUEUE_CALL_RE = re.compile(r"^(append|put|enqueue|queue|push)$", re.I)


def _is_state_expr(node: ast.AST) -> bool:
    if isinstance(node, ast.Attribute):
        return node.attr in STATE_ATTR_NAMES or bool(FLAG_NAME_RE.search(node.attr))
    if isinstance(node, ast.Name):
        return bool(FLAG_NAME_RE.search(node.id))
    return False


def _is_open_value(node: ast.AST) -> bool:
    if isinstance(node, ast.Attribute) and node.attr == "OPEN":
        return True
    if isinstance(node, ast.Constant):
        if node.value is True or node.value == 1:
            return True
        if isinstance(node.value, str) and node.value.upper() == "OPEN":
            return True
    return False


def _check_kind(test: ast.AST) -> str | None:
    if isinstance(test, ast.UnaryOp) and isinstance(test.op, ast.Not):
        return "is-not-open" if _is_state_expr(test.operand) else None

    if isinstance(test, ast.Compare) and len(test.ops) == 1:
        op = test.ops[0]
        left, right = test.left, test.comparators[0]
        matches = (_is_state_expr(left) and _is_open_value(right)) or (
            _is_state_expr(right) and _is_open_value(left)
        )
        if not matches:
            return None
        if isinstance(op, ast.Eq):
            return "is-open"
        if isinstance(op, ast.NotEq):
            return "is-not-open"
        return None

    if _is_state_expr(test):
        return "is-open"

    return None


def _has_queue_call(stmts: list[ast.stmt]) -> bool:
    for stmt in stmts:
        for n in ast.walk(stmt):
            if not isinstance(n, ast.Call):
                continue
            name = n.func.attr if isinstance(n.func, ast.Attribute) else (
                n.func.id if isinstance(n.func, ast.Name) else None
            )
            if name and QUEUE_CALL_RE.match(name):
                return True
    return False


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    out: list[dict] = []

    for node in ast.walk(tree):
        if not isinstance(node, ast.If):
            continue
        kind = _check_kind(node.test)
        if not kind:
            continue

        not_open_branch = node.body if kind == "is-not-open" else node.orelse
        if not not_open_branch:
            continue
        if _has_queue_call(not_open_branch):
            continue

        line, col, _, _ = loc(node)
        out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out

File: src/providers/openai-realtime/rules/js/buffer-audio-until-session-ready.ts
const QUEUE_CALL_NAME_PATTERN = /^(push|unshift|enqueue|queue)$/i;

/**
 * Classifies a test comparing `readyState` against OPEN: 'is-open' for
 * `readyState === OPEN`, 'is-not-open' for `readyState !== OPEN`, else null.
 */
function readyStateOpenCheckKind(test: any): 'is-open' | 'is-not-open' | null {
  if (test?.type !== 'BinaryExpression') return null;
  const isEq = test.operator === '===' || test.operator === '==';
  const isNeq = test.operator === '!==' || test.operator === '!=';
  if (!isEq && !isNeq) return null;

  const isReadyStateMember = (n: any) =>
    n?.type === 'MemberExpression' && n.property?.type === 'Identifier' && n.property.name === 'readyState';

  const isOpenValue = (n: any) =>
    (n?.type === 'MemberExpression' && n.property?.type === 'Identifier' && n.property.name === 'OPEN') ||
    (n?.type === 'Literal' && n.value === 1);

  const matches =
    (isReadyStateMember(test.left) && isOpenValue(test.right)) ||
    (isReadyStateMember(test.right) && isOpenValue(test.left));
  if (!matches) return null;
  return isEq ? 'is-open' : 'is-not-open';
}

/** True if `node` contains a call that looks like enqueueing onto a buffer (push/unshift/enqueue/queue). */
function hasQueueCall(node: any, depth = 0): boolean {
  if (!node || typeof node !== 'object' || depth > 40) return false;
  if (Array.isArray(node)) return node.some((n) => hasQueueCall(n, depth + 1));

  if (node.type === 'CallExpression') {
    const callee = node.callee;
    const calleeName =
      callee?.type === 'MemberExpression' && callee.property?.type === 'Identifier'
        ? callee.property.name
        : callee?.type === 'Identifier'
          ? callee.name
          : null;
    if (typeof calleeName === 'string' && QUEUE_CALL_NAME_PATTERN.test(calleeName)) return true;
  }

  for (const key of Object.keys(node)) {
    if (key === 'parent' || key === 'loc' || key === 'range') continue;
    const val = node[key];
    if (val && typeof val === 'object' && hasQueueCall(val, depth + 1)) return true;
  }
  return false;
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'Audio sent before an OpenAI Realtime socket is open must be buffered, not dropped',
      category: 'reliability',
      docsUrl: 'https://developers.openai.com/api/reference/resources/realtime/client-events',
      rationale:
        'Caller audio can arrive before the Realtime WebSocket finishes its connect + session.update round-trip. A readyState !== OPEN branch that only logs and returns silently drops every audio chunk that arrives in that window, meaning the first fragment of speech is lost to translation/processing on every call.',
      recommended: true,
    },
    messages: {
      audioDroppedNotBuffered:
        'Audio sent while this Realtime socket is not yet open is dropped here instead of being buffered and flushed once open.',
    },
    schema: [],
  },
  create(context: any) {
    return {
      IfStatement(node: any) {
        const kind = readyStateOpenCheckKind(node.test);
        if (!kind) return;

        // The branch taken while the socket is not yet open: the else branch
        // of `=== OPEN`, or the consequent of `!== OPEN` (early-return drop).
        const notOpenBranch = kind === 'is-open' ? node.alternate : node.consequent;
        if (!notOpenBranch) return;
        if (hasQueueCall(notOpenBranch)) return;

        context.report({ node: notOpenBranch, messageId: 'audioDroppedNotBuffered' });
      },
    };
  },
};

export const openaiRealtimeBufferAudioUntilSessionReadyRule = rule;
